[nathanielcalimag] iP

nathanielcalimag opened the pull request:
# ChatNat
> "ChatNat is one of the chatbots of all time." - Nathaniel ([source](https://www.youtube.com/watch?v=dQw4w9WgXcQ))

ChatNat frees your mind of having to remember things you need to do. It's:
- CLI-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use
- not ~that~ buggy (I think)

All you have to do is,
1. Download our ~malware~ application [here](https://github.com/nathanielcalimag/ip/releases/tag/A-Jar)
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Self-driving AI (coming soon)
- [ ] Proof for P = NP (coming soon)
___

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Duke {
    public static void main(String[] args) {
        Ui ui = new Ui();
        TaskList taskList = new TaskList();
        Parser app = new Parser();
        try {
            Storage storage = new DukeStorage();
            taskList.setStorage(storage);
            taskList.loadTasks();
        } catch (IOException | InsufficientArgumentsException |
                 DateTimeParseException | StorageCreationException e) {
            ui.displayError(e.getMessage());
        }
        ui.displayMessage(Messages.GREETING_MESSAGE, true, true);
        while (true) {
            String input = ui.getInput();
            try {
                app.executeCommand(input, taskList, ui);
            } catch (UnknownCommandException e) {
                ui.displayError(e.getMessage());
            }
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

nathanielcalimag committed:
Add greeting and exit messages

nathanielcalimag committed:
Add application structure and commands

nathanielcalimag committed:
Implement application state, insert command and list command

nathanielcalimag committed:
Implement tasks and marking functionalities

nathanielcalimag committed:
Implement deadline, event, and todo tasks

nathanielcalimag committed:
Update automated input and output text files

nathanielcalimag committed:
Implement error handling

nathanielcalimag committed:
Add deletion command

nathanielcalimag committed:
Update javadoc comments

nathanielcalimag committed:
Organise codebase into packages

All classes are stored in a flat manner under the same folder.

The lack of grouping of related classes makes the codebase messy.

Utilising packages can create structure and improve the development
process.

Let's use packages to group similar classes together. Specifically,
commands, exceptions and tasks.

nathanielcalimag committed:
Add storage functionality

nathanielcalimag committed:
Merge branch 'branch-Level-7'

nathanielcalimag committed:
Update fields of tasks to contain datetime

nathanielcalimag committed:
Merge branch 'branch-Level-8'

nathanielcalimag committed:
Implement Ui, Storage, Parser and TaskList classes

nathanielcalimag committed:
Rename main class in gradle build file

nathanielcalimag committed:
Merge branch 'branch-A-Gradle'

nathanielcalimag committed:
Rename dukeapp class to duke

nathanielcalimag committed:
Add tests for task list and parser

nathanielcalimag committed:
Merge branch 'branch-A-JUnit'

nathanielcalimag committed:
Update JavaDoc comments

nathanielcalimag committed:
Update code to meet coding standards

nathanielcalimag committed:
Implement functionality to find tasks

nathanielcalimag committed:
Merge branch 'branch-Level-9'

nathanielcalimag committed:
Merge branch 'branch-A-JavaDoc'

nathanielcalimag committed:
Merge branch 'A-CodingStandard'

nathanielcalimag committed:
Update Gradle to use Shadow

nathanielcalimag committed:
Setup CheckStyle with Gradle

nathanielcalimag committed:
Improve code standards to pass checkstyle format

nathanielcalimag committed:
Merge branch 'branch-A-CheckStyle'

SimWPEric commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,153 @@
+     * @return The deleted task's string.
+     * @throws IOException If saving tasks to file failed.
+     */
+    public String deleteTask(int index) throws IOException {
+        Task task = this.tasks.remove(index);
+        this.saveTasks();
+        return task.toString();
+    }
```
Can also throw `IndexOutOfBoundsException` in the event user attempts to delete a task at a specified index in the `List<Task> tasks` that does not exist. Same for `unmarkTask` and `markTask`.

SimWPEric commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,53 @@
+     * @param from        The starting datetime of the event.
+     * @param to          The ending datetime of the event.
+     */
+    public Event(String description, boolean isDone, LocalDateTime from, LocalDateTime to) {
+        super(description, isDone);
+        this.from = from;
+        this.to = to;
+    }
```
Is there a check to prevent `event test /from 2018-09-12 /to 2012-09-10` ?

SimWPEric reviewed (commented):
Overall, the encapsulation is impressive, and I appreciate how you've organised the classes into Java Packages. Great job! When it comes to the application's logic, I'd suggest conducting more extensive testing. For example, you can test scenarios like attempting to delete a task at index 6 from a task list containing only 2 tasks. This will help ensure the code handles such situations gracefully.

nathanielcalimag committed:
Add GUI integration

dlathyun commented on src/main/java/duke/Messages.java:
```diff
@@ -0,0 +1,40 @@
+public class Messages {
+
+    // Regular messages
+    public static final String GREETING_MESSAGE = " Hello! I'm ChatNat!\n What can I do for you?";
+
+    public static final String LIST_MESSAGE = " Here are the tasks in your list:";
+
+    public static final String FIND_MESSAGE = " Here are the matching tasks in your list:";
+
+    public static final String INSERT_MESSAGE = " Got it. I've added this task:\n   %s\n Now you "
+            + "have %d %s in the list.";
+    public static final String MARKED_MESSAGE = " Nice! I've marked this task as done:\n   %s";
+
+
+    public static final String UNMARKED_MESSAGE = " OK, I've marked this task as not done yet:\n "
```
I like it that you stored the messages in one class. Would it be better if they had common prefix? (eg: MESSAGE)

dlathyun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,266 @@
+ * Represents the Application object responsible for storing and executing commands.
+ */
+public class Parser {
+    public static final String OUTPUT_DATE_TIME_PATTERN = "d MMM yyyy, HHmm";
+    public static final String STORAGE_DATE_TIME_PATTERN = "yyyy-MM-dd HHmm";
```
I also like how you organised the pattern in this manner, but would it be better for them to have a common prefix as well? (Eg: DATE_TIME_PATTERN)

dlathyun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,266 @@
+        for (String format : dateTimeFormats) {
+            try {
+                if (format.contains("HHmm")) {
+                    return LocalDateTime.parse(dateTimeString, DateTimeFormatter.ofPattern(format));
```
Should the indentation for this part be 8 spaces more than the parent line?

dlathyun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,266 @@
+            }
+        }
+        throw new DateTimeParseException(String.format(Messages.ERROR_PREFIX, String.format(
+                Messages.INVALID_DATE_TIME_FORMAT, dateTimeString)), dateTimeString, 0);
```
Good that you kept the method name attached to the open parenthesis that follows it!

dlathyun reviewed (commented):
Overall very good in following java coding standard, could not find much issues. I like how the code is structured and organised :)

nathanielcalimag committed:
Merge branch 'branch-Level-10'

nathanielcalimag committed:
Implement assertions in the storage and parser classes

The current state of the codebase does not leverage assertions.

Using assertions in the codebase can improve the maintainability of the project and help signal to developers potential problems with their code in the future.

Let's,
* use assertions in the storage class to assert that the storage file exists before the user can use the application
* use assertions in the parser class to assert that a valid set of commands exists before the user can use the application

nathanielcalimag committed:
Refactor parser and task classes for code quality improvements

The Parser class had deeply nested code and violations in the single level of abstraction principle.

These issues can impede future code maintainence.

Extracting some of the logic in task creation and input parsing can improve readability and
code quality.

Let's,
* extract out string parsing for task creation into smaller functions
* create a helper function for date parsing to reduce nested code

nathanielcalimag committed:
Merge pull request #2 from nathanielcalimag/branch-A-Assertions

Implement assertions in the storage and parser classes

nathanielcalimag committed:
Merge branch 'master' into branch-A-CodeQuality

nathanielcalimag committed:
Merge pull request #3 from nathanielcalimag/branch-A-CodeQuality

Refactor parser and task classes for code quality improvements

nathanielcalimag committed:
Implement checks to detect duplicate tasks

nathanielcalimag committed:
Merge branch 'branch-C-DetectDuplicates'

nathanielcalimag committed:
Add a User Guide

nathanielcalimag committed:
Merge branch 'branch-A-UserGuide'

nathanielcalimag committed:
Update screenshot in user guide

nathanielcalimag committed:
Update error message formatting

nathanielcalimag committed:
Improve the styling of the dialog box

nathanielcalimag committed:
Merge branch 'branch-A-BetterGui'

nathanielcalimag committed:
Update user guide image

nathanielcalimag committed:
Improve error handling for datetime string parsing

nathanielcalimag committed:
Improve error handling on index-based commands

nathanielcalimag committed:
Update user guide to reflect true behaviour
